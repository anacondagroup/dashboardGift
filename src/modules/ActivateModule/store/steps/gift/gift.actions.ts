import { createAction } from 'redux-act';
import { TErrors } from '@alycecom/services';

import { ICustomMarketplaceSetting, IDefaultGift, IRecipientActions } from '../../activate.types';
import { GiftExchangeOptions } from '../../../constants/exchange.constants';

import { ISaveMarketplaceParams, IUpdateGiftStepPayload } from './gift.types';

const PREFIX = 'ACTIVATE_MODULE/STEPS/GIFT';

export const updateMarketplaceSettingsRequest = createAction<ISaveMarketplaceParams>(
  `${PREFIX}/UPDATE_MARKETPLACE_REQUEST`,
);
export const updateMarketplaceSettingsSuccess = createAction<ISaveMarketplaceParams>(
  `${PREFIX}/UPDATE_MARKETPLACE_SUCCESS`,
);
export const updateMarketplaceSettingsFail = createAction<TErrors>(`${PREFIX}/UPDATE_MARKETPLACE_FAIL`);

export const updateDefaultGiftRequest = createAction<{
  products: IDefaultGift[];
}>(`${PREFIX}/UPDATE_DEFAULT_GIFT_REQUEST`);
export const updateDefaultGiftSuccess = createAction<{ products: IDefaultGift[] }>(
  `${PREFIX}/UPDATE_DEFAULT_GIFT_SUCCESS`,
);
export const updateDefaultGiftFail = createAction<TErrors>(`${PREFIX}/UPDATE_DEFAULT_GIFT_FAIL`);

export const updateGiftStepRequest = createAction<IUpdateGiftStepPayload>(`${PREFIX}/UPDATE_REQUEST`);
export const updateGiftStepSuccess = createAction<{ recipientActions: IRecipientActions }>(`${PREFIX}/UPDATE_SUCCESS`);
export const updateGiftStepFail = createAction<TErrors>(`${PREFIX}/UPDATE_FAIL`);

export const updateGiftExchangeOptionsRequest = createAction<{
  giftExchangeOptions: GiftExchangeOptions | null;
}>(`${PREFIX}/UPDATE_GIFT_EXCHANGE_OPTIONS_REQUEST`);
export const updateGiftExchangeOptionsSuccess = createAction<{ giftExchangeOptions: string | null }>(
  `${PREFIX}/UPDATE_GIFT_EXCHANGE_OPTIONS_SUCCESS`,
);
export const updateGiftExchangeOptionsFail = createAction<TErrors>(`${PREFIX}/UPDATE_GIFT_EXCHANGE_OPTIONS_FAIL`);

export const updateCustomMarketplaceSetting = createAction<ICustomMarketplaceSetting>(
  `${PREFIX}/UPDATE_GIFT_CUSTOM_MARKETPLACE_REQUEST`,
);
export const updateCustomMarketplaceSettingSuccess = createAction<ICustomMarketplaceSetting>(
  `${PREFIX}/UPDATE_GIFT_CUSTOM_MARKETPLACE_SUCCESS`,
);
export const updateCustomMarketplaceSettingFail = createAction(`${PREFIX}/UPDATE_GIFT_CUSTOM_MARKETPLACE_FAIL`);

export const updateAcceptOnlyDonationSetting = createAction<number>(`${PREFIX}/UPDATE_ACCEPT_ONLY_DONATION_REQUEST`);
export const updateAcceptOnlyDonationSettingSuccess = createAction<number>(
  `${PREFIX}/UPDATE_ACCEPT_ONLY_DONATION_SUCCESS`,
);
export const updateAcceptOnlyDonationSettingFail = createAction(`${PREFIX}/UPDATE_ACCEPT_ONLY_DONATION_FAIL`);

export const setGiftExchangeOption = createAction<{ exchangeOption: GiftExchangeOptions }>(
  `${PREFIX}/SET_GIFT_EXCHANGE_OPTION`,
);

export const updateAcceptOnlyFallbackGift = createAction<IDefaultGift>(
  `${PREFIX}/UPDATE_ACCEPT_ONLY_FALLBACK_GIFT_REQUEST`,
);
export const updateAcceptOnlyFallbackGiftSuccess = createAction<IDefaultGift>(
  `${PREFIX}/UPDATE_ACCEPT_ONLY_FALLBACK_GIFT_SUCCESS`,
);
export const updateAcceptOnlyFallbackGiftFail = createAction<IDefaultGift>(
  `${PREFIX}/UPDATE_ACCEPT_ONLY_FALLBACK_GIFT_FAIL`,
);
