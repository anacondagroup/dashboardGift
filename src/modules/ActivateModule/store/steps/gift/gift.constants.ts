import { RecipientActions } from '../../activate.types';

export const RECIPIENT_ACTION_LABELS: Record<RecipientActions, string> = {
  [RecipientActions.capturePhone]: 'provide phone number',
  [RecipientActions.captureEmail]: 'provide email address',
  [RecipientActions.captureDate]: 'accept a calendar invite',
  [RecipientActions.captureQuestion]: 'answer custom questions',
  [RecipientActions.captureAffidavit]: 'accept custom terms & conditions',
  [RecipientActions.question]: '',
  [RecipientActions.affidavit]: '',
};
