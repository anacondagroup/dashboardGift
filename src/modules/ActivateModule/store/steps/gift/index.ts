export { gift } from './gift.reducer';
export type { IGiftState } from './gift.reducer';
export * from './gift.actions';
export * from './gift.selectors';
export { giftEpics } from './gift.epics';
export * from './gift.schemas';
