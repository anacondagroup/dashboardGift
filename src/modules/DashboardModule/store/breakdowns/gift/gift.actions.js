import {
  GIFT_BREAKDOWN_LOAD_REQUEST,
  GIFT_BREAKDOWN_LOAD_SUCCESS,
  GIFT_BREAKDOWN_LOAD_FAILS,
  GIFT_DOWNLOAD_REPORT_REQUEST,
  GIFT_DOWNLOAD_REPORT_SUCCESS,
  GIFT_DOWNLOAD_REPORT_FAIL,
  GIFT_DOWNLOAD_RECEIPTS_REQUEST,
  GIFT_DOWNLOAD_RECEIPTS_SUCCESS,
  GIFT_DOWNLOAD_RECEIPTS_FAIL,
  GIFT_BREAKDOWN_CLEAR,
  GIFT_BREAKDOWN_TABLE_LOAD_REQUEST,
  GIFT_BREAKDOWN_TABLE_LOAD_SUCCESS,
  GIFT_BREAKDOWN_TABLE_LOAD_FAILS,
  GIFT_BREAKDOWN_DOWNLOAD_REPORT_SUCCESS,
  GIFT_BREAKDOWN_DOWNLOAD_REPORT_FAIL,
  GIFT_BREAKDOWN_DOWNLOAD_REPORT_REQUEST,
} from './gift.types';

export const giftBreakdownRequest = options => ({
  type: GIFT_BREAKDOWN_TABLE_LOAD_REQUEST,
  payload: options,
});

export const giftBreakdownSuccess = breakdown => ({
  type: GIFT_BREAKDOWN_TABLE_LOAD_SUCCESS,
  payload: breakdown,
});

export const giftBreakdownLoadFail = error => ({
  type: GIFT_BREAKDOWN_TABLE_LOAD_FAILS,
  payload: error,
});

export const giftLoadRequest = options => ({
  type: GIFT_BREAKDOWN_LOAD_REQUEST,
  payload: options,
});

export const giftLoadSuccess = breakdown => ({
  type: GIFT_BREAKDOWN_LOAD_SUCCESS,
  payload: breakdown,
});

export const giftLoadFail = error => ({
  type: GIFT_BREAKDOWN_LOAD_FAILS,
  payload: error,
});

export const giftBreakdownDownloadReportRequest = options => ({
  type: GIFT_BREAKDOWN_DOWNLOAD_REPORT_REQUEST,
  payload: options,
});

export const giftBreakdownDownloadReportSuccess = () => ({
  type: GIFT_BREAKDOWN_DOWNLOAD_REPORT_SUCCESS,
});

export const giftBreakdownDownloadReportFail = () => ({
  type: GIFT_BREAKDOWN_DOWNLOAD_REPORT_FAIL,
});

export const giftDownloadReportRequest = options => ({
  type: GIFT_DOWNLOAD_REPORT_REQUEST,
  payload: options,
});

export const giftDownloadReportSuccess = () => ({
  type: GIFT_DOWNLOAD_REPORT_SUCCESS,
});

export const giftDownloadReportFail = () => ({
  type: GIFT_DOWNLOAD_REPORT_FAIL,
});

export const giftDownloadReceiptsRequest = options => ({
  type: GIFT_DOWNLOAD_RECEIPTS_REQUEST,
  payload: options,
});

export const giftDownloadReceiptsSuccess = () => ({
  type: GIFT_DOWNLOAD_RECEIPTS_SUCCESS,
});

export const giftDownloadReceiptsFail = () => ({
  type: GIFT_DOWNLOAD_RECEIPTS_FAIL,
});

export const giftBreakDownClear = () => ({
  type: GIFT_BREAKDOWN_CLEAR,
});
