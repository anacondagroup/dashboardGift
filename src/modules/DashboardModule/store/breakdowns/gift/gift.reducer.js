import * as R from 'ramda';

import { initialBreakdownState } from '../../../helpers/breakdownSelectors.helpers';
import { setLoading, setLoaded } from '../../../../../helpers/lens.helpers';

import {
  GIFT_BREAKDOWN_LOAD_REQUEST,
  GIFT_BREAKDOWN_LOAD_SUCCESS,
  GIFT_BREAKDOWN_LOAD_FAILS,
  GIFT_DOWNLOAD_REPORT_REQUEST,
  GIFT_DOWNLOAD_REPORT_SUCCESS,
  GIFT_DOWNLOAD_REPORT_FAIL,
  GIFT_DOWNLOAD_RECEIPTS_REQUEST,
  GIFT_DOWNLOAD_RECEIPTS_SUCCESS,
  GIFT_DOWNLOAD_RECEIPTS_FAIL,
  GIFT_BREAKDOWN_CLEAR,
  GIFT_BREAKDOWN_TABLE_LOAD_REQUEST,
  GIFT_BREAKDOWN_TABLE_LOAD_SUCCESS,
  GIFT_BREAKDOWN_TABLE_LOAD_FAILS,
  GIFT_BREAKDOWN_DOWNLOAD_REPORT_REQUEST,
  GIFT_BREAKDOWN_DOWNLOAD_REPORT_SUCCESS,
  GIFT_BREAKDOWN_DOWNLOAD_REPORT_FAIL,
} from './gift.types';

export const initialState = {
  ...initialBreakdownState,
  isReceiptsLoading: false,
};

export const mapBreakdown = gift =>
  gift && (gift.batchName || gift.sentProduct)
    ? gift
    : {
        id: gift.id || '',
        recipientId: gift.recipient_id,
        fullName: gift.recipient_full_name || '',
        company: gift.recipient_company || '',
        gift: gift.product ? gift.product.name : '',
        campaign: gift.campaign || '',
        campaignId: gift.campaign_id,
        sentBy: gift.sent_as || '',
        method: gift.gift_invitation_type || '',
        sentOn: gift.sent_on || '',
        giftStatus: gift.status.name || '',
        giftStatusId: gift.status.id || '',
        canChooseOptions: gift.can_choose_options,
        teamId: gift.team_id,
        scheduledAt: gift.scheduled_at || null,
      };

export const reducer = (state = initialState, action) => {
  switch (action.type) {
    case GIFT_DOWNLOAD_RECEIPTS_REQUEST:
      return {
        ...state,
        isReceiptsLoading: true,
      };
    case GIFT_DOWNLOAD_RECEIPTS_FAIL:
    case GIFT_DOWNLOAD_RECEIPTS_SUCCESS:
      return {
        ...state,
        isReceiptsLoading: false,
      };
    case GIFT_BREAKDOWN_TABLE_LOAD_REQUEST:
    case GIFT_BREAKDOWN_LOAD_REQUEST:
      return {
        ...state,
        ...setLoading(true, state),
      };

    case GIFT_BREAKDOWN_TABLE_LOAD_SUCCESS:
    case GIFT_BREAKDOWN_LOAD_SUCCESS:
      return {
        ...state,
        ...R.compose(setLoading(false), setLoaded(true))(state),
        breakdown: action.payload.gifts.map(mapBreakdown),
        pagination: action.payload.meta.pagination,
      };

    case GIFT_BREAKDOWN_TABLE_LOAD_FAILS:
    case GIFT_BREAKDOWN_LOAD_FAILS:
      return {
        ...state,
        ...setLoading(false, state),
        error: action.payload,
      };
    case GIFT_BREAKDOWN_DOWNLOAD_REPORT_REQUEST:
    case GIFT_DOWNLOAD_REPORT_REQUEST:
      return {
        ...state,
        isReportLoading: true,
      };
    case GIFT_BREAKDOWN_DOWNLOAD_REPORT_SUCCESS:
    case GIFT_DOWNLOAD_REPORT_SUCCESS:
    case GIFT_BREAKDOWN_DOWNLOAD_REPORT_FAIL:
    case GIFT_DOWNLOAD_REPORT_FAIL:
      return {
        ...state,
        isReportLoading: false,
      };
    case GIFT_BREAKDOWN_CLEAR:
      return {
        ...initialState,
      };
    default:
      return state;
  }
};
