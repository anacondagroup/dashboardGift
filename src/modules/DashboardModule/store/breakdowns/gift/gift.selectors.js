import * as R from 'ramda';

import { getBreakdownsState } from '../breakdowns.selectors';
import { makeGetTeamById } from '../../../../../store/teams/teams.selectors';
import { makeGetMemberById } from '../../members/members.selectors';
import { createSelectors } from '../../../helpers/breakdownSelectors.helpers';
import { makeGetCampaignById } from '../../../../../store/campaigns/campaigns.selectors';

export const [
  getGiftState,
  getGiftBreakdown,
  getGiftIsLoading,
  getGiftReportIsLoading,
  getGiftIsLoaded,
  getGiftPagination,
] = createSelectors(getBreakdownsState)('gift');

export const getReportFileName = apiParams => state => {
  let team = null;
  let date = '';
  let member = null;
  let campaign = null;

  if (apiParams.memberId) {
    const { fullName: memberName } = makeGetMemberById(apiParams.memberId)(state);
    member = `-${memberName}`;
  }

  if (apiParams.teamId) {
    const { name: teamName } = makeGetTeamById(apiParams.teamId)(state);
    team = `-${teamName}-Team`;
  }

  if (apiParams.dateRangeFrom) {
    date = `-${apiParams.dateRangeFrom}`;
  }

  if (apiParams.dateRangeTo) {
    date += `-${apiParams.dateRangeTo}`;
  }

  if (apiParams.campaignId) {
    const { name: campaignName } = makeGetCampaignById(apiParams.campaignId)(state);
    campaign = `-${campaignName}-Campaign`;
  }

  return R.replace(/\./g, '_', `Gifts breakdown${team || ''}${campaign || ''}${member || ''}${date || ''}`).concat(
    '.xlsx',
  );
};

export const searchGifts = (search = '') =>
  R.filter(
    R.compose(
      R.any(value => value.toString().toLowerCase().includes(search.toString().toLowerCase())),
      R.values,
    ),
  );

export const getSortBy = R.cond([
  [R.equals('name'), R.always(R.compose(R.toLower, item => `${item.firstName} ${item.lastName}`))],
  [R.equals('giftStatus'), R.always(R.prop('giftStatusId'))],
  [R.equals('sentOn'), R.always(R.compose(sentOn => (sentOn ? new Date(sentOn).getTime() : 0), R.prop('sentOn')))],
  [R.T, field => R.compose(R.toLower, R.prop(field))],
]);

export const sortGifts = (order, orderBy) =>
  R.compose(orderBy === 'asc' ? R.identity : R.reverse, R.sortBy(getSortBy(order)));
