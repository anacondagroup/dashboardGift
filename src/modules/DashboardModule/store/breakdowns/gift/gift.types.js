export const GIFT_BREAKDOWN_LOAD_REQUEST = 'GIFT_BREAKDOWN_LOAD_REQUEST';
export const GIFT_BREAKDOWN_LOAD_SUCCESS = 'GIFT_BREAKDOWN_LOAD_SUCCESS';
export const GIFT_BREAKDOWN_LOAD_FAILS = 'GIFT_BREAKDOWN_LOAD_FAILS';

export const GIFT_BREAKDOWN_TABLE_LOAD_REQUEST = 'GIFT_BREAKDOWN_TABLE_LOAD_REQUEST';
export const GIFT_BREAKDOWN_TABLE_LOAD_SUCCESS = 'GIFT_BREAKDOWN_TABLE_LOAD_SUCCESS';
export const GIFT_BREAKDOWN_TABLE_LOAD_FAILS = 'GIFT_BREAKDOWN_TABLE_LOAD_FAILS';

export const GIFT_DOWNLOAD_REPORT_REQUEST = 'GIFT_DOWNLOAD_REPORT_REQUEST';
export const GIFT_DOWNLOAD_REPORT_SUCCESS = 'GIFT_DOWNLOAD_REPORT_SUCCESS';
export const GIFT_DOWNLOAD_REPORT_FAIL = 'GIFT_DOWNLOAD_REPORT_FAIL';

export const GIFT_BREAKDOWN_DOWNLOAD_REPORT_REQUEST = 'GIFT_BREAKDOWN_DOWNLOAD_REPORT_REQUEST';
export const GIFT_BREAKDOWN_DOWNLOAD_REPORT_SUCCESS = 'GIFT_BREAKDOWN_DOWNLOAD_REPORT_SUCCESS';
export const GIFT_BREAKDOWN_DOWNLOAD_REPORT_FAIL = 'GIFT_BREAKDOWN_DOWNLOAD_REPORT_FAIL';

export const GIFT_DOWNLOAD_RECEIPTS_REQUEST = 'GIFT_DOWNLOAD_RECEIPTS_REQUEST';
export const GIFT_DOWNLOAD_RECEIPTS_SUCCESS = 'GIFT_DOWNLOAD_RECEIPTS_SUCCESS';
export const GIFT_DOWNLOAD_RECEIPTS_FAIL = 'GIFT_DOWNLOAD_RECEIPTS_FAIL';

export const GIFT_BREAKDOWN_CLEAR = 'GIFT_BREAKDOWN_CLEAR';
