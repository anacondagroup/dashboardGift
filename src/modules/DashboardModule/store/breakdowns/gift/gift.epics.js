import { ofType } from 'redux-observable';
import { catchError, debounce, map, mergeMap, takeUntil, tap, withLatestFrom } from 'rxjs/operators';
import { of, timer } from 'rxjs';

import { getApiCallQuery, urlBuilder } from '../../../../../helpers/url.helpers';

import {
  GIFT_BREAKDOWN_DOWNLOAD_REPORT_REQUEST,
  GIFT_BREAKDOWN_LOAD_REQUEST,
  GIFT_BREAKDOWN_TABLE_LOAD_REQUEST,
  GIFT_DOWNLOAD_RECEIPTS_REQUEST,
  GIFT_DOWNLOAD_REPORT_REQUEST,
} from './gift.types';
import {
  giftBreakdownDownloadReportFail,
  giftBreakdownDownloadReportSuccess,
  giftBreakdownLoadFail,
  giftBreakdownSuccess,
  giftDownloadReceiptsFail,
  giftDownloadReceiptsSuccess,
  giftDownloadReportFail,
  giftDownloadReportSuccess,
  giftLoadFail,
  giftLoadSuccess,
} from './gift.actions';
import { getReportFileName } from './gift.selectors';

export const breakdownAPIUrl = options => `/enterprise/dashboard/breakdown/gifts?${urlBuilder(options)}`;
export const breakdownReportUrl = options => `/enterprise/report/gift_breakdown?${urlBuilder(options)}`;

export const buildApiUrl = options => `/enterprise/dashboard/breakdown/gifts?${getApiCallQuery(options)}`;
export const buildReportUrl = options => `/enterprise/report/gift_breakdown?${getApiCallQuery(options)}`;
export const buildReceiptUrl = options => `/enterprise/report/gift_receipts?${getApiCallQuery(options)}`;

export const loadBreakdownEpic = (action$, state$, { apiService }) =>
  action$.pipe(
    ofType(GIFT_BREAKDOWN_TABLE_LOAD_REQUEST),
    debounce(() => timer(300)),
    map(action => breakdownAPIUrl(action.payload)),
    mergeMap(url =>
      apiService.get(url).pipe(
        map(response => giftBreakdownSuccess(response)),
        catchError(response => of(giftBreakdownLoadFail(response))),
        takeUntil(action$.ofType(GIFT_BREAKDOWN_TABLE_LOAD_REQUEST)),
      ),
    ),
  );

export const downloadGiftBreakdownReportEpic = (action$, state$, { apiService, downloadFile }) =>
  action$.pipe(
    ofType(GIFT_BREAKDOWN_DOWNLOAD_REPORT_REQUEST),
    map(action => [breakdownReportUrl(action.payload), action]),
    mergeMap(([url, action]) =>
      apiService.getFile(url).pipe(
        withLatestFrom(state$.pipe(map(getReportFileName(action.payload)))),
        tap(([blob, filename]) => {
          downloadFile(blob, filename);
        }),
        map(() => giftBreakdownDownloadReportSuccess()),
        catchError(() => [giftBreakdownDownloadReportFail()]),
      ),
    ),
  );

export const loadGiftBreakdownEpic = (action$, state$, { apiService }) =>
  action$.pipe(
    ofType(GIFT_BREAKDOWN_LOAD_REQUEST),
    debounce(() => timer(300)),
    map(action => buildApiUrl(action.payload)),
    mergeMap(url =>
      apiService.get(url).pipe(
        map(response => giftLoadSuccess(response)),
        catchError(response => of(giftLoadFail(response))),
        takeUntil(action$.ofType(GIFT_BREAKDOWN_LOAD_REQUEST)),
      ),
    ),
  );

export const downloadReportEpic = (action$, state$, { apiService, downloadFile }) =>
  action$.pipe(
    ofType(GIFT_DOWNLOAD_REPORT_REQUEST),
    map(action => [buildReportUrl(action.payload), action]),
    mergeMap(([url, action]) =>
      apiService.getFile(url).pipe(
        withLatestFrom(state$.pipe(map(getReportFileName(action.payload)))),
        tap(([blob, filename]) => {
          downloadFile(blob, filename);
        }),
        map(() => giftDownloadReportSuccess()),
        catchError(() => [giftDownloadReportFail()]),
      ),
    ),
  );

export const downloadReceiptsEpic = (action$, state$, { apiService, downloadFile }) =>
  action$.pipe(
    ofType(GIFT_DOWNLOAD_RECEIPTS_REQUEST),
    map(action => [buildReceiptUrl(action.payload), action]),
    mergeMap(([url]) =>
      apiService.getFile(url).pipe(
        withLatestFrom(state$.pipe(map(() => 'Gifts receipt'))),
        tap(([blob, filename]) => {
          downloadFile(blob, filename);
        }),
        map(() => giftDownloadReceiptsSuccess()),
        catchError(() => [giftDownloadReceiptsFail()]),
      ),
    ),
  );

export const giftEpic = [
  loadBreakdownEpic,
  loadGiftBreakdownEpic,
  downloadGiftBreakdownReportEpic,
  downloadReportEpic,
  downloadReceiptsEpic,
];
