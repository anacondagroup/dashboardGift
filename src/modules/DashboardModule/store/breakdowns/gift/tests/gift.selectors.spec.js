import { sortGifts, getReportFileName } from '../gift.selectors';
import { fullStateMock } from '../../../../../../mocks/fullState.mock';
import { buildApiUrl, buildReportUrl } from '../gift.epics';

describe('gift selector', () => {
  describe('getReportFileName', () => {
    it('Should build name without params', () => {
      const actual = getReportFileName({})(fullStateMock);
      expect(actual).toMatchSnapshot();
    });

    it('Should build name with all params', () => {
      const actual = getReportFileName({
        teamId: '1',
        dateRangeFrom: '2019-03-12',
        dateRangeTo: '2019-03-13',
        campaignId: '31',
        memberId: '1210',
      })(fullStateMock);

      expect(actual).toMatchSnapshot();
    });
  });
  describe('buildApiUrl', () => {
    it('Should build url', () => {
      const actual = buildApiUrl({
        sort: 'name',
        sortDirection: 'asc',
        search: 'asd',
      });

      expect(actual).toMatchSnapshot();
    });
  });
  describe('buildReportUrl', () => {
    it('Should build url', () => {
      const actual = buildReportUrl({
        sort: 'name',
        sortDirection: 'asc',
        search: 'asd',
      });

      expect(actual).toMatchSnapshot();
    });
  });
  describe('sortGifts', () => {
    it('Should sort by status id when sort column is giftStatus', () => {
      const gifts = [
        {
          id: 1,
          giftStatus: 'status',
          giftStatusId: 1,
        },
        {
          id: 1,
          giftStatus: 'status',
          giftStatusId: 5,
        },
        {
          id: 1,
          giftStatus: 'status',
          giftStatusId: 1,
        },
        {
          id: 1,
          giftStatus: 'status',
          giftStatusId: 2,
        },
        {
          id: 1,
          giftStatus: 'status',
          giftStatusId: 1,
        },
        {
          id: 1,
          giftStatus: 'status',
          giftStatusId: 3,
        },
      ];

      expect(sortGifts('giftStatus', 'asc')(gifts)).toEqual([
        {
          id: 1,
          giftStatus: 'status',
          giftStatusId: 1,
        },
        {
          id: 1,
          giftStatus: 'status',
          giftStatusId: 1,
        },
        {
          id: 1,
          giftStatus: 'status',
          giftStatusId: 1,
        },
        {
          id: 1,
          giftStatus: 'status',
          giftStatusId: 2,
        },
        {
          id: 1,
          giftStatus: 'status',
          giftStatusId: 3,
        },
        {
          id: 1,
          giftStatus: 'status',
          giftStatusId: 5,
        },
      ]);

      expect(sortGifts('giftStatus', 'desc')(gifts)).toEqual([
        {
          id: 1,
          giftStatus: 'status',
          giftStatusId: 5,
        },
        {
          id: 1,
          giftStatus: 'status',
          giftStatusId: 3,
        },
        {
          id: 1,
          giftStatus: 'status',
          giftStatusId: 2,
        },
        {
          id: 1,
          giftStatus: 'status',
          giftStatusId: 1,
        },
        {
          id: 1,
          giftStatus: 'status',
          giftStatusId: 1,
        },
        {
          id: 1,
          giftStatus: 'status',
          giftStatusId: 1,
        },
      ]);
    });

    it('Should sort by full name', () => {
      const gifts = [
        {
          id: 1,
          firstName: 'alyce bob',
        },
        {
          id: 1,
          firstName: 'alyce alyce',
        },
        {
          id: 1,
          firstName: 'bob alyce',
        },
        {
          id: 1,
          firstName: 'bob bob',
        },
      ];

      expect(sortGifts('giftStatus', 'desc')(gifts)).toEqual([
        {
          id: 1,
          firstName: 'bob bob',
        },
        {
          id: 1,
          firstName: 'bob alyce',
        },
        {
          id: 1,
          firstName: 'alyce alyce',
        },
        {
          id: 1,
          firstName: 'alyce bob',
        },
      ]);

      expect(sortGifts('giftStatus', 'asc')(gifts)).toEqual([
        {
          id: 1,
          firstName: 'alyce bob',
        },
        {
          id: 1,
          firstName: 'alyce alyce',
        },
        {
          id: 1,
          firstName: 'bob alyce',
        },
        {
          id: 1,
          firstName: 'bob bob',
        },
      ]);
    });

    it('Should sort by sentOn', () => {
      const gifts = [
        {
          id: 1,
          firstName: 'alyce bob',
          sentOn: '12/14/18',
        },
        {
          id: 1,
          firstName: 'alyce alyce',
          sentOn: '12/14/19',
        },
        {
          id: 1,
          firstName: 'alyce alyce',
        },
        {
          id: 1,
          firstName: 'bob alyce',
          sentOn: '12/14/13',
        },
        {
          id: 1,
          firstName: 'bob bob',
          sentOn: '01/14/18',
        },
      ];

      expect(sortGifts('sentOn', 'asc')(gifts)).toEqual([
        {
          id: 1,
          firstName: 'alyce alyce',
        },
        {
          id: 1,
          firstName: 'bob alyce',
          sentOn: '12/14/13',
        },
        {
          id: 1,
          firstName: 'bob bob',
          sentOn: '01/14/18',
        },
        {
          id: 1,
          firstName: 'alyce bob',
          sentOn: '12/14/18',
        },
        {
          id: 1,
          firstName: 'alyce alyce',
          sentOn: '12/14/19',
        },
      ]);

      expect(sortGifts('sentOn', 'desc')(gifts)).toEqual([
        {
          id: 1,
          firstName: 'alyce alyce',
          sentOn: '12/14/19',
        },
        {
          id: 1,
          firstName: 'alyce bob',
          sentOn: '12/14/18',
        },
        {
          id: 1,
          firstName: 'bob bob',
          sentOn: '01/14/18',
        },
        {
          id: 1,
          firstName: 'bob alyce',
          sentOn: '12/14/13',
        },
        {
          id: 1,
          firstName: 'alyce alyce',
        },
      ]);
    });
  });
});
