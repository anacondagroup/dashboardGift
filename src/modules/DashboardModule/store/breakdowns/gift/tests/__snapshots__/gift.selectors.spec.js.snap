// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gift selector buildApiUrl Should build url 1`] = `"/enterprise/dashboard/breakdown/gifts?page=1&per_page=10&search=asd&sort=recipient_full_name&sort_direction=asc"`;

exports[`gift selector buildReportUrl Should build url 1`] = `"/enterprise/report/gift_breakdown?page=1&per_page=10&search=asd&sort=recipient_full_name&sort_direction=asc"`;

exports[`gift selector getReportFileName Should build name with all params 1`] = `"Gifts breakdown-WEX: sales dep-Team-campaign 4 test-Campaign-Shyanne Borer-2019-03-12-2019-03-13.xlsx"`;

exports[`gift selector getReportFileName Should build name without params 1`] = `"Gifts breakdown.xlsx"`;
